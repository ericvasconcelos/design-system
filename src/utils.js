export { arrayCombinations } from 'utils/arrayCombinations'
export { getBackgroundColor } from 'utils/getBackgroundColor'
export { getBorder } from 'utils/getBorder'
export { getButtonStyle } from 'utils/getButtonStyle'
export { getBorderRadius } from 'utils/getBorderRadius'
export { getColorFromName } from 'utils/getColorFromName'
export { getGuttersNames } from 'utils/getGuttersNames'
export { getIconNames } from 'utils/getIconNames'
export { getPadding } from 'utils/getPadding'
export { getPaddingNames } from 'utils/getPaddingNames'
export { getTypography } from 'utils/getTypography'
export { parsePalleteName } from 'utils/parsePalleteName'
export { styleColorValidator } from 'utils/styleColorValidator'
export { useAriaProps } from 'utils/useAriaProps'

export { DateUtils } from 'utils/DateUtils'
export { QueryUtils } from 'utils/QueryUtils'
export { ObjectUtils }  from 'utils/ObjectUtils'
export { CookiesUtils } from 'utils/CookiesUtils'
export { MoneyUtils } from 'utils/MoneyUtils'
export { NumberUtils } from 'utils/NumberUtils'