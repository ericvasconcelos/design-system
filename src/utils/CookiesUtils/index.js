export class CookiesUtils {
  static toObject(string = '') {
    const cookies = {}
    if (string !== '') {
      const parts = string.split('; ')
      parts.forEach(part => {
        const [key, value] = part.split('=')
        cookies[key] = value
      })
    }
    return cookies
  }
}
