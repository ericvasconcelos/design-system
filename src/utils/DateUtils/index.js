export class DateUtils {
  static toDate(string, format = 'pt-br') {
    let day = ''
    let month = ''
    let year = ''

    switch (format) {
      case 'pt-br':
        [day, month, year] = string?.split('/')
        break
      case 'iso':
        [year, month, day] = string?.split('-')
        break
      default:
        throw new Error('Format invalid')
    }

    return new Date(year, parseInt(month, 10) - 1, day)
  }

  static getAge(date) {
    if (!(date instanceof Date)) {
      throw new TypeError('Date invalid')
    }

    const difference = Date.now() - date.getTime()

    return Math.abs(new Date(difference).getUTCFullYear() - 1970)
  }

  static toString(date, format = 'pt-br') {
    if (!(date instanceof Date)) {
      throw new TypeError('Date invalid')
    }

    const day = date.getDate() < 10 ? `0${date.getDate()}` : date.getDate()
    const month = date.getMonth() + 1 < 10 ? `0${date.getMonth() + 1}` : date.getMonth() + 1
    const year = date.getFullYear()

    switch (format) {
      case 'pt-br':
        return `${day}/${month}/${year}`
      case 'iso':
        return `${year}-${month}-${day}`
      default:
        throw new Error('Format invalid')
    }
  }

  static validate(date, format = 'pt-br') {
    const monthDaysMap = {
      '01': 31,
      '02': 29,
      '03': 31,
      '04': 30,
      '05': 31,
      '06': 30,
      '07': 31,
      '08': 31,
      '09': 30,
      10: 31,
      11: 30,
      12: 31
    }
    let isValid = false

    let day = ''
    let month = ''
    let year = ''

    switch (format) {
      case 'pt-br':
        [day, month, year] = date?.split('/')
        break
      case 'iso':
        [year, month, day] = date?.split('-')
        break
      default:
        throw new Error('Format invalid')
    }

    isValid = isValid || parseInt(day, 10) <= monthDaysMap[month]
    isValid = isValid && parseInt(year, 10) > 1900

    return isValid
  }

  static convertMonthToYear(month) {
    return parseInt(month, 10) / 12
  }

  static formatPeriod(period, periodType) {
    const periodTypeMap = {
      monthly: { singular: 'mÃªs', plural: 'meses' },
      yearly: { singular: 'ano', plural: 'anos' },
      daily: { singular: 'dia', plural: 'dias' }
    }
    return `${period} ${periodTypeMap[periodType][period < 2 ? 'singular' : 'plural']}`
  }
}
