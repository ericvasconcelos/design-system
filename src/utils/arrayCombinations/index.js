export const arrayCombinations = (...args) => {
  const returnable = []
  const max = args.length - 1
  function helper(arr, i) {
    for (let j = 0, { length } = args[i]; j < length; j++) {
      const newArray = arr.slice(0) // clone arr
      newArray.push(args[i][j])
      if (i === max) returnable.push(newArray)
      else helper(newArray, i + 1)
    }
  }
  helper([], 0)
  return returnable
}
