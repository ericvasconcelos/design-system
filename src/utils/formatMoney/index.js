import { toMoney, toNumber } from 'vanilla-masker'

export const formatMoney = entry => {
  const moneyFormatter = {
    precision: 2,
    separator: ',',
    delimiter: '.',
    unit: 'R$'
  }

  return toMoney(toNumber(entry), moneyFormatter)
}
