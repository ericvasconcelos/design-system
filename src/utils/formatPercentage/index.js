import { toMoney, toNumber } from 'vanilla-masker'

export const formatPercentage = entry => {
  const percentageFormatter = {
    precision: 1,
    separator: ',',
    suffixUnit: '%'
  }
  return toMoney(toNumber(entry), percentageFormatter)
}
