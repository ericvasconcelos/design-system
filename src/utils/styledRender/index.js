import React from 'react'

import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { theme } from 'styles/theme'

export const styledRender = (children, options = {}) => {
  const rendered = render(<ThemeProvider theme={theme}>{children}</ThemeProvider>, options)
  return {
    ...rendered,
    rerender: (ui, rerenderOptions = {}) =>
      styledRender(ui, { container: rendered.container, ...rerenderOptions })
  }
}
