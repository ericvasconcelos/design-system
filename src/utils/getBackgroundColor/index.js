import { getColorFromName } from 'utils/getColorFromName'

export const getBackgroundColor = ({ theme, styleColor }) =>
  `background-color: ${getColorFromName({ theme, styleColor })};`
