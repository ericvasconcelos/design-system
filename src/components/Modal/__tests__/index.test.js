import React from 'react'

import { act, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { styledRender } from 'utils/styledRender'

import { Modal } from '../index.js'

jest.useFakeTimers()
describe('Modal should render the component correctly when', () => {
  const titleText = 'Eu proident veniam aliqua voluptate id.'
  const handleClose = jest.fn()
  handleClose.mockName('handleClose')
  const contentText = 'Anim eiusmod nisi ex cillum adipisicing sint laborum.'

  it('show is true', () => {
    const { getByTestId, getByText } = styledRender(
      <Modal title={titleText} id="test" onClose={handleClose} show>
        {contentText}
      </Modal>
    )
    act(() => {
      jest.runAllTimers()
    })
    const component = getByTestId('Modal')
    const title = getByText(titleText)
    const content = getByText(contentText)
    expect(component).toBeInTheDocument()
    expect(title).toBeInTheDocument()
    expect(content).toBeInTheDocument()

    expect(component).toMatchSnapshot()
  })

  it('closes when overlay is clicked', async () => {
    const { getByTestId, rerender } = styledRender(
      <Modal title={titleText} id="test" onClose={handleClose} show>
        {contentText}
      </Modal>
    )
    handleClose.mockImplementationOnce(() => {
      rerender(
        <Modal title={titleText} id="test" onClose={handleClose} show={false}>
          {contentText}
        </Modal>
      )
    })

    const component = getByTestId('Modal')

    act(() => {
      jest.runAllTimers()
      fireEvent.transitionEnd(component)
    })

    expect(component).toBeInTheDocument()

    act(() => {
      userEvent.click(component)
      jest.runAllTimers()
      fireEvent.transitionEnd(component)
    })

    expect(handleClose).toHaveBeenCalled()
    expect(component).not.toBeInTheDocument()
  })

  it('closes when ESC  is pressed', async () => {
    const { container, getByTestId, rerender } = styledRender(
      <Modal title={titleText} id="test" onClose={handleClose} show>
        {contentText}
      </Modal>
    )

    handleClose.mockImplementationOnce(() => {
      rerender(
        <Modal title={titleText} id="test" onClose={handleClose} show={false}>
          {contentText}
        </Modal>
      )
    })
    act(() => {
      jest.runAllTimers()
    })

    const component = getByTestId('Modal')
    expect(component).toBeInTheDocument()

    act(() => {
      userEvent.type(container, '{esc}')
      jest.runAllTimers()
      fireEvent.transitionEnd(component)
    })

    expect(handleClose).toHaveBeenCalled()
    expect(component).not.toBeInTheDocument()
  })

  it('do nothing when wrapper is clicked', async () => {
    const { getByTestId } = styledRender(
      <Modal title={titleText} id="test" onClose={handleClose} show>
        {contentText}
      </Modal>
    )

    act(() => {
      jest.runAllTimers()
    })
    const component = getByTestId('Modal')

    const modalWrapper = component.querySelector('[aria-describedby=\'title\']')
    act(() => {
      userEvent.click(modalWrapper)
    })
    expect(handleClose).not.toHaveBeenCalled()
  })

  it('do nothing when any other key is presse', async () => {
    const { container, getByTestId } = styledRender(
      <Modal title={titleText} id="test" onClose={handleClose} show>
        {contentText}
      </Modal>
    )

    act(() => {
      jest.runAllTimers()
    })
    const component = getByTestId('Modal')

    act(() => {
      userEvent.type(container, '{enter}')
    })

    act(() => {
      userEvent.type(
        container,
        'Pariatur non est nostrud non est tempor ipsum aliqua ullamco magna excepteur et sit'
      )
    })

    const modalWrapper = component.querySelector('#test')

    act(() => {
      userEvent.type(modalWrapper, '{enter}')
    })

    expect(component).toBeInTheDocument()
    expect(handleClose).not.toHaveBeenCalled()
  })
})
