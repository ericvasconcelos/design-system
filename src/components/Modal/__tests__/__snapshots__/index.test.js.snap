// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render the component correctly when show is true 1`] = `
.c2 {
  color: #00224B;
  font-size: 1.5rem;
  font-weight: 500;
  line-height: 1.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-letter-spacing: -0.02rem;
  -moz-letter-spacing: -0.02rem;
  -ms-letter-spacing: -0.02rem;
  letter-spacing: -0.02rem;
  font-family: 'DM Sans',sans-serif;
}

.c0 {
  opacity: 1;
  background-color: #000000e0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  z-index: 10;
}

.c0:focus {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c3 {
  border-top: #A6B5BB solid 1px;
  width: 100%;
  height: 1px;
  margin: 0;
}

.c1 {
  width: 100%;
  background-color: #FFFFFF;
  max-width: 550px;
  padding: 2rem 1.5rem;
  border-radius: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 1.5rem;
}

<div
  class="c0"
  data-testid="Modal"
  role="dialog"
>
  <div
    aria-describedby="title"
    class="c1"
    id="test"
  >
    <h2
      class="c2"
      data-testid="Title"
      id="modal-title"
    >
      Eu proident veniam aliqua voluptate id.
    </h2>
    <hr
      class="c3"
    />
    <div
      class=""
    >
      Anim eiusmod nisi ex cillum adipisicing sint laborum.
    </div>
  </div>
</div>
`;
